ramips: Refactor lzma-loader

antoniovazquezblanco opened the pull request:
This is my first contribution to the project. Sorry if I miss something or something is not up to standards.

This PR aims to simplify the LZMA loader in ramips target:
- Removed unused mk files.
- Unified cache options into the main makefile.
- Unified board code into a weak definition that can be overriden if needed.
- Fixed UART addresses in multiple platforms.

antoniovazquezblanco committed:
ramips: lzma-loader: Refactor loader

* Delete unused lantiq makefile
* Delete redundant makefiles and unify them into the main makefile
* Refactor and unify board code into a single file
* Add support and review subtarget specific board support

Signed-off-by: Antonio Vázquez <antoniovazquezblanco@gmail.com>

mrkiko commented:
Thanks a lot for your contribution!
And - especially since it's the first, congrats for the topic you choose.
This work es extremely valuable.
You probably will be asked to detail more about what you've done in your commits - especially the one regarding board support refactoring.


That said, if you feel brave enough, you _may_ also try to unify the different lzma-loader implementations.

antoniovazquezblanco commented:
I think it may be viable to unify them all and I think I may know how to deal with the different target/subtarget needs. That being said, I may need some pointers on where to place the code and I will need help with testing...

The testing part is easy: I do not have devices on which to try this and I may be able to setup emulation for many of them but certainly I would need testing from users...

Where to place the code: Currently the loaders reside under each arch target. Do we want to keep lzma loader code in the main tree or should it be moved to a separate repo? In any case, do you have any pointers on where to place it? I may be able to progressively migrate devices to the new codebase starting with the ramips arch to make this manageable...

Thanks for the pointer on the details of my last commit. Looking back I have not been paying attention to my commit msgs.

Thanks!

Spudz76 commented:
I wouldn't mind some selectable other compression schemes.  Mostly ZSTD.  I started trying to implement but have no idea what I'm doing.  Mostly a decompressor implementation that does not use libraries so it works in early boot / kernel mode.

antoniovazquezblanco commented:
I like the idea. Indeed, I want to use this for a router that requires a kernel compressed in lzma so compressing the kernel twice using the same algorithm does not yield great results...

On the other hand, I would prefer to do this in small steps if possible. Migrating to a new place and allowing broader device support is kind of a challenge for me given this is my first contribution without any new features for the time being.

Thanks for the feedback!

blocktrron commented on target/linux/ramips/image/lzma-loader/src/board.c:
```diff
@@ -0,0 +1,37 @@
+extern uint32_t volatile * const UART_LSR;
+extern uint32_t const UART_LSR_THRE_MASK;
+
+// Generic initialization function. Can be overwritten in board-$(SUBTARGET).c
+// files if needed.
+void __attribute__((weak)) board_init(void)
+{
+}
+
+// Generic putc function. Can be overwritten in board-$(SUBTARGET).c files if
+// needed.
```
I'd prefer a consistent usage of C style comments here.

antoniovazquezblanco replied:
Easy! Will do! :)

antoniovazquezblanco replied:
Done!

antoniovazquezblanco replied:
Should I resolve this comments or should @blocktrron? What is your usual way to proceed? Thanks!

blocktrron commented on target/linux/ramips/image/lzma-loader/src/board-rt305x.c:
```diff
@@ -0,0 +1,16 @@
+
+#include <stdint.h>
+
+uint32_t volatile * const UART_BASE = (uint32_t volatile * const) 0x10000500;
+uint32_t volatile * const UART_THR = UART_BASE + 0x04;
+uint32_t volatile * const UART_LSR = UART_BASE + 0x1c;
+uint32_t const UART_LSR_THRE_MASK = 0x20;
```
I don't like keeping the pointer declaration seperated from the source-file.

I do prefer having a single header file which uses compiler-definitions for the UART base address. All platforms emulate a 16850, so the register offsets and function are identical.

antoniovazquezblanco replied:
I also don't like C file declaration of pointers... It is plainly not nice but I did it for a couple of reasons:

* SOC definitions were not available in header files unless I filled the makefiles with conditionals... I don't know if this was due to the fact that the project has a makefile to call another makefile... Will do some testing.
* Not all UARTs have the same offsets. It seems MT7621 breaks the pattern... :/

I will perfom some tests and will propose a better solution, preferably based on header definitions :).

antoniovazquezblanco replied:
From what I have read, `CONFIG_` variables will not be available in lzma-loader unless passed via make calls. Thinking about this I was wondering what would be the ideal scenario and thought of the following:

What is the relevant information I need for making the UART work?
I need the UART peripheral "model" and the address where it is placed. I believe that passing that information would be a saner option than passing a `CONFIG_SOC_XXXX` macro and deriving the UART behaviour and address from that information.

I guess it may be kind of difficult to come up with UART peripheral "models"... 

I could provide a dummy (silent) default UART configuration. If an UART model is specified via a definition then corresponding C files will get compiled to provide the corresponding functionality. Peripheral base address could then be specified via another paramter.

Would this approach fit your concerns? Any cons regarding this?

Thanks!

oliv3r replied:
If there's any way you can use kernel config variables, I actually tackled the same problem https://gitlab.com/olliver/openwrt/openwrt/-/commit/23a975a79f1a4b4c4f19e552343b23061497737e :) (i'll upstream it once I'm confident in it)

antoniovazquezblanco replied:
@oliv3r, your suggestion is to pass down the `CONFIG_SOC_XXX` variable from the kernel config down to the loader via Make calls? Is there a better way to let those variables be known to the loader Makefiles and source code?

oliv3r replied:
I'd be surprised if that's possible or acceptable :) Just some food for thought is all.

dhewg replied:
Drive by comment, but all that sounds like what linux already does? grep for DEBUG_UART_

antoniovazquezblanco replied:
Nice, those CONFIG_DEBUG_UART_ are really useful. A combination of `CONFIG_SOC_XXXX`, `CONFIG_DEBUG_UART_BASE` and optionally `CONFIG_DEBUG_UART_CLOCK` from the kernel config should be enough for the job...

Now onto how to get those in the loader code from the Kernel config...
Thanks!

antoniovazquezblanco replied:
Currently using SUBTARGET Makefile variable. I hope this will do for the time being.

antoniovazquezblanco replied:
Should I resolve this comments or should @blocktrron? What is your usual way to proceed? Thanks!

dhewg commented:
While you're at it, I noticed this by coincidence:
```
-fwhole-program
[snip]
           This option should not be used in combination with -flto.  Instead
           relying on a linker plugin should provide safer and more precise
           information.
```
and
```
target/linux/ath79/image/lzma-loader/src/Makefile:LDFLAGS       += -flto -fwhole-program
target/linux/lantiq/image/lzma-loader/src/Makefile:LDFLAGS      += -flto -fwhole-program
target/linux/ramips/image/lzma-loader/src/Makefile:LDFLAGS      += -flto -fwhole-program -Wl,-z,max-page-size=4096
```

...

So I guess it should be like this instead;
```
package/utils/busybox/Makefile:TARGET_LDFLAGS += -flto=jobserver -fuse-linker-plugin
```

Does that work (better)? Does it affect the file size?

mpratt14 commented:
> I think it may be viable to unify them all

something I would like to see, so that work on the loaders main code can apply to all targets at the same time.
something I would like to do, if I didn't have so much else to do 😆 



> Where to place the code: Currently the loaders reside under each arch target. Do we want to keep lzma loader code in the main tree or should it be moved to a separate repo?

i think the answer to that is pretty simple, common code in `target/linux/generic`, and arch specific code in each arch folder

Spudz76 commented:
Also, this breaks for me at runtime testing so I assume something is wrong.  Not much useful output just the uboot jumps to kernel and then nothing happens (hang).  I do not have JTAG to break and inspect any further once the serial console freezes.

Maybe I have to manually nuke all the lzma-loader build dir(s) and stuff to ensure clean rebuild from the patched sources, it could be from hot-swapping code and the build system doesn't notice.

mpratt14 commented:
> Maybe I have to manually nuke all the lzma-loader build dir(s) and stuff to ensure clean rebuild from the patched sources

no, it was recently set to do that during a build

see 
a01d23e755ba46f41e667d558d82d4871d7f5450
1bff0752bd5e7feb0f311224a09b3bf217a9aeb3

oliv3r commented:
@antoniovazquezblanco great contribution, especially the unification part. I also have a custom bootloader (realtek) which requires the payload (the kernel) to be LZMA compressed. I really do not want to add 'yet a nother LZMA wrapper' so am super excited to hopefully make use of yours when ready!

antoniovazquezblanco commented:
Sorry as I kind of been away for some days. Nice to see so much interaction here. I will need some time to chew on this and will get back to it...

antoniovazquezblanco commented:
> While you're at it, I noticed this by coincidence:
> 
> ```
> -fwhole-program
> [snip]
>            This option should not be used in combination with -flto.  Instead
>            relying on a linker plugin should provide safer and more precise
>            information.
> ```
> 
> and
> 
> ```
> target/linux/ath79/image/lzma-loader/src/Makefile:LDFLAGS       += -flto -fwhole-program
> target/linux/lantiq/image/lzma-loader/src/Makefile:LDFLAGS      += -flto -fwhole-program
> target/linux/ramips/image/lzma-loader/src/Makefile:LDFLAGS      += -flto -fwhole-program -Wl,-z,max-page-size=4096
> ```
> 
> ...
> 
> So I guess it should be like this instead;
> 
> ```
> package/utils/busybox/Makefile:TARGET_LDFLAGS += -flto=jobserver -fuse-linker-plugin
> ```
> 
> Does that work (better)? Does it affect the file size?

Thanks @dhewg for the comment. I am looking into this...

From https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html:
```
-fwhole-program

    Assume that the current compilation unit represents the whole program being compiled. All public
functions and variables with the exception of main and those merged by attribute externally_visible 
become static functions and in effect are optimized more aggressively by interprocedural optimizers.

    With -flto this option has a limited use. In most cases the precise list of symbols used or exported
from the binary is known the resolution info passed to the link-time optimizer by the linker plugin. It is
still useful if no linker plugin is used or during incremental link step when final code is produced (with
-flto -flinker-output=nolto-rel).


-flto[=n]

...

 If the program does not require any symbols to be exported, it is possible to combine -flto and
-fwhole-program to allow the interprocedural optimizers to use more aggressive assumptions which
may lead to improved optimization opportunities. Use of -fwhole-program is not needed when
linker plugin is active (see -fuse-linker-plugin). 

...

Use -flto=auto to use GNU make’s job server, if available, or otherwise fall back to autodetection of
the number of CPU threads present in your system. 
```

Given we are compiling final code, the compiled code represent the whole program being compiled, that we want to optimize it and that the program does not export any simbol I think that the usage of `-flto` and `-fwhole-program` provides agressive optimization without problems... Could you please post a reference for your quote to further read on this?

Regarding the `-flto=jobserver` option, from the quote above, it may be interesting to use `-flto=auto` instead wich tries the `jobserver` strat and if it does not work, it attempts to detect processor count by other means. What do you think?

Thanks again for the feedback! :)

antoniovazquezblanco commented:
> Also, this breaks for me at runtime testing so I assume something is wrong. Not much useful output just the uboot jumps to kernel and then nothing happens (hang). I do not have JTAG to break and inspect any further once the serial console freezes.

Could you please tell me for what device/SOC does it break?
There was specific code for some SOCs in the original implementation that seems was not being used so maybe it was not needed... But I may have introduced some new bugs ;)

Thank you @Spudz76!

antoniovazquezblanco commented:
While working on the loader I also have found that there is a Makefile just to call make on an `src` subfolder that contains another Makefile and the real loader project. It seems this pattern is used in some other places around OpenWRT tree.

Can someone point why this is done? Should I simplify the tree and only use one Makefile? Is there any reason why I shouldn't?

Thanks!

Spudz76 commented:
Usually that pattern is to retain the upstream Makefile in whatever condition it's in, with the upper Makefile to adapt things to the OpenWRT tree.  Since lzma-loader doesn't really have an upstream (final edition) I don't think there would be much problem unifying it now.  Keep original author credits with it somewhere, add yours.

antoniovazquezblanco commented:
Did a review of the PR. I would appreaciate some new feedback. I hope I did not miss anything. Thanks!

Spudz76 commented on target/linux/ramips/image/lzma-loader/src/Makefile:
```diff
@@ -54,7 +53,27 @@ LDFLAGS		+= -flto -fwhole-program -Wl,-z,max-page-size=4096
+endif
+
+ifeq ($(SUBTARGET),mt7621)
+CFLAGS	+= -DSOC_MT621
```
Should be `-DSOC_MT7621`

antoniovazquezblanco replied:
Fixed, I have just realized that CI only builds the kernel. Will compile all the targets and will report back if something fails.

Thanks!

antoniovazquezblanco replied:
Should I resolve this comments or should @Spudz76? What is your usual way to proceed? Thanks!

Spudz76 replied:
I can't mark resolved because I don't have that button.

mpratt14 commented:
I think the BOARD variable should be preserved, even though device specific features of the loader are not specifically needed right now in ramips, they might be in the future

reference commit from not long ago in ath79

e9335c292003e15901ef5f665750f9493132eb28

antoniovazquezblanco commented:
> I think the BOARD variable should be preserved, even though device specific features of the loader are not specifically needed right now in ramips, they might be in the future
> 
> reference commit from not long ago in ath79
> 
> [e9335c2](https://github.com/openwrt/openwrt/commit/e9335c292003e15901ef5f665750f9493132eb28)

I would like to keep making PRs to unify different implementations of the loader in the future. I would vote to remove it as it is not needed and to include it in the future if needed with the required refactor. As far as I know there are multiple other non used variables and I would prefer to remove them also because they require maintenance and make it less clear...

If you, @mpratt14, are positive that you want them in place, I will put it back in place :)

mpratt14 commented:
if you add it back, it might be in a different file, so I'm not sure

I'm being picky of course, but in my opinion removing a variable from the makefile is out of scope of your commit which has the goal of moving and consolidating makefile content to the C code, if I understand correctly

I think the maintainers like to see the smallest diff possible that achieves the goals, I've been trying to adapt to that method of development as well.

I know they want the makefiles for the loader between different targets to be as similar as possible (there have been other commits to make them "in sync")

also don't forget to squash the 2 commits together, and if you need help with git just ask :)

antoniovazquezblanco commented:
Rebased with latest master branch. Reinstated the BOARD variable to keep things as simple as possible. Fixed a compilation issue. Compiled all ramips targets for testing.

Thanks!

oliv3r commented on target/linux/ramips/image/lzma-loader/Makefile:
```diff
@@ -11,13 +11,14 @@ include $(TOPDIR)/rules.mk
 LZMA_TEXT_START	:= 0x80a00000
 LOADER		:= loader.bin
 LOADER_NAME	:= $(basename $(notdir $(LOADER)))
-LOADER_DATA 	:=
```
a small nit; (unrelated) commits are preferred in their own commits, so there's nice atomic commits :)

oliv3r commented on target/linux/ramips/image/lzma-loader/src/Makefile:
```diff
@@ -23,16 +23,16 @@ FLASH_START	:=
-CACHE_FLAGS	:=
+SUBTARGET	:=
+CACHE_FLAGS	:= -DCONFIG_ICACHE_SIZE="(32 * 1024)" \
+		   -DCONFIG_DCACHE_SIZE="(16 * 1024)" -DCONFIG_CACHELINE_SIZE=32
```
While you are only working on 'step 1' of the unification (just for ramips, where I'm hoping step 2 will be that it will be generic enough for openwrt as a whole) I wonder if this is the best place fort his. Can we not get (and otherwise set) this in a makefile higher? idk

antoniovazquezblanco replied:
Moved up a Makefile level

oliv3r commented:
I like where this is going, looking good so far!

antoniovazquezblanco commented:
> I like where this is going, looking good so far!

Thank you for your feedback! :)

Spudz76 commented:
New crash for me:
```
touch /media/scratch/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linksys_ea7300-v1-kernel.bin.src/.prepared
make -C /media/scratch/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linksys_ea7300-v1-kernel.bin.src CROSS_COMPILE="mipsel-openwrt-linux-musl-" \
        LZMA_TEXT_START=0x81800000 \
        LOADER_DATA=/media/scratch/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linksys_ea7300-v1-kernel.bin \
        FLASH_START= \
        FLASH_OFFS= \
        FLASH_MAX= \
        BOARD="MT7621 based boards" \
        PLATFORM="mt7621" \
        SUBTARGET="mt7621" \
        CACHE_FLAGS="-DCONFIG_ICACHE_SIZE="(32 * 1024)" -DCONFIG_DCACHE_SIZE="(16 * 1024)" -DCONFIG_CACHELINE_SIZE=32" \
        clean all
bash: -c: line 9: syntax error near unexpected token `('
bash: -c: line 9: `     CACHE_FLAGS="-DCONFIG_ICACHE_SIZE="(32 * 1024)" -DCONFIG_DCACHE_SIZE="(16 * 1024)" -DCONFIG_CACHELINE_SIZE=32" \'
make[6]: *** [Makefile:44: loader-compile] Error 1
make[6]: Leaving directory '/media/scratch/openwrt/target/linux/ramips/image/lzma-loader'
```

Reapplied patch from here, no idea what is different from when it did work.

Spudz76 commented:
This fixes it (simply escaped the offending doublequotes in Makefile)
```
diff --git a/target/linux/ramips/image/lzma-loader/Makefile b/target/linux/ramips/image/lzma-loader/Makefile
index e52a107b13..31a8e90197 100644
--- a/target/linux/ramips/image/lzma-loader/Makefile
+++ b/target/linux/ramips/image/lzma-loader/Makefile
@@ -19,8 +19,8 @@ FLASH_MAX     :=
 BOARD          :=
 PLATFORM       :=
 SUBTARGET      :=
-CACHE_FLAGS    := -DCONFIG_ICACHE_SIZE="(32 * 1024)" \
-                  -DCONFIG_DCACHE_SIZE="(16 * 1024)" -DCONFIG_CACHELINE_SIZE=32
+CACHE_FLAGS    := -DCONFIG_ICACHE_SIZE=\"(32 * 1024)\" \
+                  -DCONFIG_DCACHE_SIZE=\"(16 * 1024)\" -DCONFIG_CACHELINE_SIZE=32

 ifeq ($(TARGET_DIR),)
 TARGET_DIR     := $(KDIR)
```

mpratt14 commented:
can we have the resulting number instead of math?

mpratt14 commented on target/linux/ramips/image/lzma-loader/src/board.c:
```diff
@@ -0,0 +1,55 @@
+#include <stdint.h>
+
+#if defined(SOC_MT7620) || defined(SOC_RT3883)
+#define UART_BASE			0xb0000c00
+#define UART_THR			(UART_BASE + 0x0004)
+#define UART_LSR			(UART_BASE + 0x001c)
+#define UART_LSR_THRE_MASK	(1 << 6)
```
now getting very picky, can we get rid of the leading zeros and bitwise shift like in the defines for other SOC

`1 << 6` is `0x40`

antoniovazquezblanco replied:
Done!

antoniovazquezblanco commented:
Addresed comments from @Spudz76 and @mpratt14 .
Also rebased with the latest master.

mpratt14 commented:
but do we need the math for `16 * 1024` or can we just simply put the number, so then we don't need quotes at all

antoniovazquezblanco commented:
> but do we need the math for `16 * 1024` or can we just simply put the number, so then we don't need quotes at all

I don't think so but AFAIK those values are not even used... I would vote to remove `CONFIG_ICACHE_SIZE` and `CONFIG_DCACHE_SIZE`...

mpratt14 commented:
good point, I can only find references of it being defined but not used...

mpratt14 commented on target/linux/ramips/image/lzma-loader/src/Makefile:
```diff
@@ -54,7 +53,27 @@ LDFLAGS		+= -flto -fwhole-program -Wl,-z,max-page-size=4096
+$(error "Please specify a SUBTARGET!")
+endif
+
+ifeq ($(strip $(SUBTARGET)),mt7620)
+CFLAGS	+= -DSOC_MT7620
+endif
+
+ifeq ($(SUBTARGET),mt7621)
+CFLAGS	+= -DSOC_MT7621
+endif
+
+ifeq ($(SUBTARGET),rt305x)
+CFLAGS	+= -DSOC_RT305X
+endif
+
+ifeq ($(SUBTARGET),rt3883)
+CFLAGS	+= -DSOC_RT3883
+endif
```
let's use strip in all of them to remain consistent

antoniovazquezblanco replied:
Done!

antoniovazquezblanco commented:
I forgot to comment after the latest fix. Should have addressed everything.

Thanks

mpratt14 commented:
I think this is ready

however I think this comment no longer applies?
`Can be overwritten in board-$(SUBTARGET).c files if needed.`

A different version of the function would now be done with preprocessor statements right?

antoniovazquezblanco commented:
Removed the comments altoguether. It seems self-explanatory enought to my eyes. Thanks!

antoniovazquezblanco commented:
Tanks everyone for helping me improve my code. I hope next time will be less work for you.

Leo-PL commented:
@antoniovazquezblanco 
I have to dig out this old PR, because I found that it causes a regression, that stops ZTE MF283+ from booting. The board would stop right after decompressing the kernel. before the initial LZMA loader prompt would appear - reverting c31319b66934f554f6c337b4cfa888b1fb74faa4 fixes the issue.

I don't want to submit a revert just yet, because MF283+ needs other fixes too, but I'd be glad if you could double-check if the subtarget configuration options weren't improperly changed in the process.

mpratt14 commented:
@Leo-PL @hauke fix is here (but please test it) #14151 

unfortunate that none of us caught this earlier...

MIPS SOCs remap memory for various reasons, like so RAM can be at 0x0 or something like that

Honestly, it's hard to find information on it... other than just reading code, but I linked a very nice doc

Leo-PL commented:
@mpratt14 will test right away, thanks!
